plugin/backend/ipvariant.d

```diff
@@ -0,0 +1,906 @@
+        .each!((a) {a.setKind(ClassType.Gmock); r.put(a); });
+    // dfmt on
+
+    return r;
+}
+
+/** Translate namspaces and the content to test double implementations.
+ */
+Nullable!CppNamespace translate(CppNamespace input, ref Container container,
+        Controller ctrl, Parameters params) {
+    import std.algorithm : map, filter, each, uniq;
+    import std.array : empty, array;
+    import cpptooling.data.representation : CppNs;
+    import cpptooling.generator.adapter : makeAdapter, makeSingleton;
+    import cpptooling.generator.func : makeFuncInterface;
+    import cpptooling.generator.gmock : makeGmock;
+
+    static auto makeGmockInNs(CppClass c, Parameters params) {
+        import cpptooling.data.representation : CppNs;
+
+        auto ns = CppNamespace.make(CppNs(cast(string) params.getMainNs));
+        ns.setKind(NamespaceType.TestDouble);
+        ns.put(makeGmock!ClassType(c));
+        return ns;
+    }
+
+    auto ns = input.dup;
+    //dfmt off
+    input.namespaceRange
+        .map!(a => translate(a, container, ctrl, params))
+        .filter!(a => !a.isNull)
+        .each!(a => ns.put(a.get));
+    
+    input.classRange
+        .each!(a => ns.put(a));
+    // dfmt on
+
+    Nullable!CppNamespace rval;
+    if (!ns.namespaceRange.empty || !ns.classRange.empty) {
+        rval = ns;
+    }
+
+    return rval;
+}
+
+/** Translate the structure to code.
+ *
+ * order is important, affects code layout:
+ *  - anonymouse for test double instance.
+ *  - implementations using double.
+ *  - adapter.
+ */
+void generate(CppRoot r, Controller ctrl, Parameters params,
+        Generator.Modules modules, ref const(Container) container)
+in {
+    import std.array : empty;
+
+    assert(r.funcRange.empty);
+}
+body {
+    import std.algorithm : each, filter;
+    import std.array;
+    import std.typecons;
+    import cpptooling.data.symbol.types : USRType;
+    import cpptooling.generator.func : generateFuncImpl;
+    import cpptooling.generator.gmock : generateGmock;
+    import cpptooling.generator.includes : generateIncludes;
+    import cpptooling.data.representation;
+    import cpptooling.analyzer.type;
+
+    generateIncludes(ctrl, params, modules.hdr);
+
+    static void gmockGlobal(T)(T r, CppModule gmock, Parameters params) {
+        foreach (a; r.filter!(a => cast(ClassType) a.kind == ClassType.Gmock)) {
+            generateGmock(a, gmock, params);
+        }
+    }
+
+    // recursive to handle nested namespaces.
+    // the singleton ns must be the first code generate or the impl can't
+    // use the instance.
+    @trusted static void eachNs(LookupT)(CppNamespace ns, Parameters params,
+            Generator.Modules modules, CppModule impl_singleton, LookupT lookup, ref string[] cifaces) {
+        import std.variant;
+        import std.algorithm : canFind;	
+	    string currnsrp;	
+        string currns = ns.fullyQualifiedName;
+
+        auto currns_spl = (cast(string)(currns)).split("::");
```

Comment:
Probably not, old code that hasn't been removed since I got resideInNs() and fullyQualifiedName to work properly.